internal/controllers/core/podlogstream/podsource.go

```diff
@@ -72,45 +80,101 @@ func (s *PodSource) TearDown() {
 // Register the pods for this stream.
 //
 // Set up any watches we need.
-func (s *PodSource) handleReconcileRequest(ctx context.Context, name types.NamespacedName, pls *PodLogStream) {
+func (s *PodSource) handleReconcileRequest(plsNN types.NamespacedName, pls *PodLogStream) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
-	s.indexer.OnReconcile(name, pls)
+	s.indexer.OnReconcile(plsNN, pls)
 
-	ns := pls.Spec.Namespace
+	var ns string
+	if pls != nil {
+		ns = pls.Spec.Namespace
+	}
+
+	// ensure namespace watch registered if PLS object exists + has valid spec
 	if ns != "" {
-		_, ok := s.watchesByNamespace[ns]
+		pw, ok := s.watchesByNamespace[ns]
 		if !ok {
-			ctx, cancel := context.WithCancel(ctx)
-			pw := podWatch{ctx: ctx, cancel: cancel, namespace: ns}
+			pw = newPodWatch(s.ctx, ns)
 			s.watchesByNamespace[ns] = pw
-			go s.doWatch(pw)
+			go s.doWatchWithRetry(pw)
+		}
+		pw.watchers[plsNN] = true
+	}
+
+	// clean up any pre-existing namespace watches
+	// this covers both PLS changing namespace in spec + PLS deletion
+	for nsKey, watch := range s.watchesByNamespace {
+		if watch.watchers[plsNN] && nsKey != ns {
+			delete(watch.watchers, plsNN)
+			s.maybeCleanupWatch(nsKey)
 		}
 	}
 }
 
-// Process pod events and make sure they trigger a reconcile.
-func (s *PodSource) doWatch(pw podWatch) {
-	defer pw.cancel()
+// maybeCleanupWatch removes a namespace watch if there are no more active watchers.
+func (s *PodSource) maybeCleanupWatch(nsKey string) {
```

Comment:
i'm a little bit worried about the flow:
- user restarts a pod
- pod is deleted
- watches are deleted
- pods are created
- watches are recreated

and wonder if we should have some sort of debouncer on the cleanup (e.g., only debounce if it's been unused for 10s)

maybe this is a premature optimization though and it's not worth the extra complexity.

wdyt?

Reply:
I had the same thought - I can add that, I don't think it'll be a ton of extra code.

(I think in most cases you won't end up hitting that because you'll still have at least one other PLS from a different resource holding the namespace watch open, but it still probably merits avoiding it if possible)